### ISSUE TITLE
Support collection rename

### ISSUE THREAD

**Issue by jingkl** [milvus-io/milvus#4812]
The user wants the collection rename API to switch between two collections
Implementation steps:
1. User creates a new collection B 
2. User uses the collection rename API to change the name of collection B to collection A.
3. User delete the original collection A

API Reference:

Collection_Rename(collection_name ,collection_rename)

Parameters

collection_rename(Str)- The name of the new collection

Previous issue Link:https://github.com/milvus-io/milvus/issues/1714

**Comment by jingkl** [comment]
After a survey of users, collection alias and collection rename both serve to switch between two collections https://github.com/milvus-io/milvus/issues/4328,If there are other comments, please add them in this issue

**Comment by yhmo** [comment]
Duplicate with #1714 
This feature need to change meta format, which may lead to large scale code change, so we didn't implement.
In the [v1.0 announcement](https://github.com/milvus-io/milvus/discussions/4800), we say we will have a new milvus with new architect, the features that not implemented in v1.0 will be considered/designed in the new milvus.

**Comment by lsgrep** [comment]
Hi, I would like to take a shot at implementing the `collection rename` feature if it is OK with you guys. I am still reading through the codebase. I have some rough ideas at the moment, but I would love to post refined design ideas next week and discuss them with you guys before starting the actual coding.

**Comment by yhmo** [comment]
> Hi, I would like to take a shot at implementing the `collection rename` feature if it is OK with you guys. I am still reading through the codebase. I have some rough ideas at the moment, but I would love to post refined design ideas next week and discuss them with you guys before starting the actual coding.

Thank you for willing to help us design/implement this feature. 
Are you reading the v2.0 codebase(in the master branch)? We marked this feature as "2.0-Backlog" now.
Before we starting a new feature, we put design doc in this site: https://github.com/milvus-io/milvus/tree/master/docs/design_docs, if you have a raw design/idea, please feel free to create a new design doc in this site for discussion.

**Comment by lsgrep** [comment]
Yes, I am reading the `master` branch code.  OK, I will follow your design process & open an issue for detailed discussion soon.  Thank you.

**Comment by yhmo** [comment]
> Yes, I am reading the `master` branch code. OK, I will follow your design process & open an issue for detailed discussion soon. Thank you.

Since this issue #4812 is for this feature "rename collection", we don't need create another issue, just add discussion/design in this issue.

**Comment by lsgrep** [comment]
Hi, sorry for the late reply.  Here is my rough design. 

### CollectionAlias Proposal

#### Background 
<img width="1408" alt="Screen Shot 2021-07-22 at 7 00 07 PM" src="https://user-images.githubusercontent.com/3893940/126629152-588d90a0-96d8-428e-a783-7ac99b3e8b16.png">

This is my rough understanding of the boundary between collection name & collection id within `Milvus` for the lifecycle of a user request. (This is not accurate since `Proxy` also caches collection `name` <=> `id` mapping.)

The `CollectionID` is used everywhere and it is kinda difficult to update `CollectionName` <=> `CollectionID` mapping globally.  In order to keep the changes minimum, we can introduce the `CollectionAlias` abstraction. 

####  Key `CollectionAlias` Properties
1. As the name indicates, it is an alias to an existing collection. 
2. The collection alias can be updated to a new collection.
3. Within `RootCoordinator`, `Proxy` and all the key components, `CollectionName` and `CollectionAlias` are equal.
   e.g. `MetaTable.GetCollectionByName(collectionName string, ts typeutil.Timestamp)`  can receive `CollectionAlias` and return corresponding `CollectionInfo`
4. `CollectionAlias` ∧ `CollectionName` = ∅  . `CollectionAlias` cannot collide with existing `CollectionName`. 


#### `Rename Collection` with `CollectionAlias`
With `CollectionAlias` we can implement rename collection functionally  as such:
1.  Create Collection `A` & insert data. 
2. Create CollectionAlias `Z` to `A`.
3. Create Collection `B` & insert data.
4. Update CollectionAlias `Z` to `B`. 
5. Users can access new data with the `Z` alias.

**Comment by yhmo** [comment]
@lsgrep 
Could you provide a definition of the `CollectionAlias` class? Please list its members, interfaces.
And some interfaces may be changed in the RootCoor/Proxy/MetaTable, please list new interfaces or modifications to show how the CollectionAlias is used by the RootCoor/Proxy/MetaTable.

I don't understand the steps 1 ~ 5, are these steps to rename a collection "A" to "B"?

**Comment by yhmo** [comment]
After more detail is confirmed, we will put the raw proposal into this page: https://wiki.lfaidata.foundation/display/MIL/MEP+10+--+Support+rename+collection
Some other developers may give their concern and suggestion. Coding work can be started once there is no concern. If you have made code change on your local, you also can give the commit link on your repository to help developers understand your solution.

**Comment by lsgrep** [comment]
Hi @yhmo ,

Running steps 1~5,  would not change collection A or B.  We just updated the alias from `Z -> A` to `Z -> B`.

We are trying to solve the problem that `Collection Rename` achieves, but without actually renaming the collection. Maybe it is appropriate to call this `MEP 10 -- Support Collection Alias`.

The reason we are not considering implementing physical collection rename are twofold:
- Implementation complexity 
  - persistent & in-flight states has to be updated.
  e.g. we have to update all the `CollectionInfo` instances. 
  ```protobuf
   message CollectionInfo {
    int64 ID = 1;
    schema.CollectionSchema schema = 2;
    uint64 create_time = 3;
    repeated int64 partitionIDs = 4;
    repeated string partitionNames = 5;
    repeated FieldIndexInfo field_indexes = 6;
    repeated string virtual_channel_names = 7;
    repeated string physical_channel_names = 8;
    repeated uint64 partition_created_timestamps = 9;
  }
  ```
- Potential performance considerations
  - Even if we have implemented physical renaming, it is not difficult to see that Physical Collection Rename will be a heavy operation, and there might be considerable downtime during rename phase.

On the other hand, using CollectionAlias will be somewhat lightweight, 
and problem surface will be much smaller, and existing state & access patterns won't be affected during or after the CollectionAlias operations.

# CollectionAlias

```go
type CollectionAlias string
```
I am unsure whether it is necessary to create a new type, we can discuss further during implementation phase, assuming you approve this proposal first.


## User facing API changes
There will be new user facing APIs.
- CreateAlias
  - Adds new entry in the `metaTable`, RootCoordinator
- DropAlias
  - Removes entry from the `metaTable`, RootCoordinator
  - Removes entries from `metaCache`, Proxy
- UpdateAlias
  - Updates entry in the `metaTable`, RootCoordinator
  - Update entries in the `metaCache`, Proxy


## Potential changes in RootCoordinator


#### Data
I presumed `CollectionAlias` will  be type `string` and MetaTable will have to store extra mapping between `CollectionAlias` <=> `CollectionID`.
e.g.   There will be new `collAlias2ID` field in `metaTable` implementation.

```go
type metaTable struct {
	collName2ID     map[string]typeutil.UniqueID               // collection name to collection id
    // NEW Addition
	collAlias2ID    map[string]typeutil.UniqueID               // collection alias to collection id
}
```

#### Access Patterns
1. Internally `MetaTable.GetCollectionByName()` can receive CollectionAlias and return `*pb.CollectionInfo`
```go
func (mt *metaTable) GetCollectionByName(collectionName string, ts typeutil.Timestamp) (*pb.CollectionInfo, error) {mt.ddLock.RLock()
  defer mt.ddLock.RUnlock()
  if ts == 0 {
  vid, ok := mt.collName2ID[collectionName]
  if !ok {
    if vid, ok = mt.collAlias2ID[collectionName]; !ok{
        return nil, fmt.Errorf("can't find collection: " + collectionName)
    }
  }
  ...
}


```
2. User can pass CollectionAlias as CollectionName for query.

### References
- https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html

**Comment by xiaofan-luan** [comment]
Very Impressive design,  Lgtm. 
Few Quick questions:
1) Are we supporting multiple alias to 1 collection？
2)Once we do alias, can we still operate on collection name? for instance, if we alias collection a to bb, should we still load a/release a or load bb, release bb?
3) when search happens, what is the path? we first check if it is a collection name or alias name? then each search need a extra etcd read? do we want to cache the alias at proxy to avoid extra etcd read?

**Comment by xiaofan-luan** [comment]
from the definition MetaTable.GetCollectionByName, collAlias2ID seems to be a in memory structure. How we update the map if create alias or delete alias happens?

**Comment by lsgrep** [comment]
Hi @xiaofan-luan , 
1. Yes.  Users can create multiple collection alias that refers to the same collection. 
2. CollectionAlias won't be intrusive. All the preexisting access patterns won't be affected. Users can use the collection name without problems and users can do collection release/load with collection alias.
3. There won't be extra `etcd` access since we have `metaTable.collAlias2ID` in the memory. 
4. `Create/Drop/Alter` alias requests eventually call corresponding GRPC endpoints in RootCoordinator. Update happens within those methods.

**Comment by xiaofan-luan** [comment]
> Hi @xiaofan-luan ,
> 
> 1. Yes.  Users can create multiple collection aliases that refers to the same collection.
> 2. CollectionAlias won't be intrusive. All the preexisting access patterns won't be affected. Users can use the collection name without problems and users can do collection release/load with collection alias.
> 3. There won't be extra `etcd` access since we have `metaTable.collAlias2ID` in the memory.  Yes.  `Proxy` will cache the `alias` as the way it caches collection name.
> 4. `Create/Drop/Alter` alias requests eventually call corresponding GRPC endpoints in RootCoordinator. `collAlias2ID ` updates happens within those methods.

Hey lsgrep, how is the process for now? do we need extra help here?

**Comment by lsgrep** [comment]
Hi @xiaofan-luan , Thanks for reaching out. I am working on the implementation hopefully I will submit PR next week.
